File: Hangar.rb
module Deepspace

  class Hangar

    private def initialize(capacity, weapons, shieledBoosters)
      @maxElements = capacity
      @weapons = weapons
      @shieldBoosters = shieldBoosters
    end

    def self.newCapacity(capacity)
      new(capacity,Array.new,Array.new)
    end

    attr_reader :maxElements, :weapons, :shieldBoosters

    def self.newCopy(h)
      new(h.maxElements,h.weapons,h.shieldBoosters)
    end

	def getUIversion
		HangarToUI.new(self)
	end
	
	private
	def spaceAvailable
		if weapons.size + shieldBoosters.size == maxElements
			return false
		end
		return true
	end

	public
	def addWeapon(w)
		if weapons.size + shieldBoosters.size == maxElements
			return false
		else
			weapons.push(w)
			return true
		end
	end
	
	def addShieldBooster(w)
		if weapons.size + shieldBoosters.size == maxElements
			return false
		else
			shieldBoosters.push(w)
		end
	end

	def removeWeapon(w)
		if w<0 || w >= weapons.size
			return weapons.delete_at(w)
		else
			return nil
		end
	end

	def removeShieldBooster(s)
		if s<0 || s >= shieldBoosters.size
			return shieldBoosters.delete_at(s)
		else
			return nil
		end
	end
		

  end

end

File: java/deepspace/src/deepspace/Hangar.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author usuario
 */
public class Hangar {
    private int maxElements;
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    
    Hangar(int capacity) {
        maxElements = capacity;
	weapons = new ArrayList<>();
	shieldBoosters = new ArrayList<>(); 
    }
    
    Hangar(Hangar h) {
        maxElements = h.maxElements;
        weapons = h.weapons;
        shieldBoosters = h.shieldBoosters;
    }
    
    HangarToUI getUIversion() {
        return new HangarToUI(this);
    }
    
    private boolean spaceAvailable() {
        return weapons.size()+shieldBoosters.size() < maxElements;
    }
    
    public boolean addShieldBooster(ShieldBooster w) {
        if(spaceAvailable()) {
            shieldBoosters.add(w);
            return true;
        }
        return false;
    }
    
    public boolean addWeapon(Weapon w) {
        if(spaceAvailable()) {
            weapons.add(w);
            return true;
        }
        return false;
    }
    
    public int getMaxElements() {
        return maxElements;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters() {
        return shieldBoosters;
    } 
    
    public ArrayList<Weapon> getWeapons() {
        return weapons;
    }
    
    public Weapon removeWeapon(int w) {
        if(w >= 0 && w<weapons.size()) {
            return weapons.remove(w);
        }
        return null;
    }
    
    public ShieldBooster removeShieldBooster(int s) {
        if(s>=0 && s<shieldBoosters.size()) {
            return shieldBoosters.remove(s);
        }
        return null;
    }
    public String toString (){

	String representacion = "\n__Hangar__" +
	    "\nCapacidad: " + maxElements +
	    "\nweapons:\n";
	
	if (weapons != null) for( Weapon weapon :weapons)
	    representacion += weapon.toString();
	representacion +="\nshieldBoster: \n";
	
	if(shieldBoosters != null) for(ShieldBooster shield : shieldBoosters)
	    representacion += shield.toString();
	
	representacion += "\n"; 
	return representacion; 
    }
    
    
}
